graphix/command.py

```diff
@@ -3,65 +3,92 @@
 from __future__ import annotations
 
 import abc
-import enum
-from typing import TYPE_CHECKING
+import dataclasses
+from abc import ABC
+from enum import Enum
+from typing import Any, Literal
 
 import numpy as np
-from pydantic import BaseModel, ConfigDict
+import typing_extensions
 
 from graphix import clifford
 from graphix.pauli import Pauli, Plane, Sign
 from graphix.states import BasicStates, State
 
-if TYPE_CHECKING:
+if typing_extensions.TYPE_CHECKING:
     from graphix.clifford import Clifford
 
 Node = int
 
 
-class CommandKind(enum.Enum):
+class CommandKind(Enum):
     N = "N"
     M = "M"
     E = "E"
     C = "C"
     X = "X"
     Z = "Z"
-    T = "T"
     S = "S"
 
 
-class Command(BaseModel, abc.ABC):
+class Command(ABC):
```

Comment:
Tagged unions do not seem to be recognized by `mypy` if their union type is a base class. I would rename `Command` as `BaseCommand` and define a type alias `Command = N | E | M | X | Z | C | T`, that `mypy` understands.

We can see that tagged unions using a base class as union type are not recognized by `mypy` on the following minimal example:
```python
"""Test tagged union."""

from __future__ import annotations

import dataclasses
import typing
from dataclasses import dataclass
from enum import Enum
from typing import Literal


class Tag(Enum):
    """Tag enum."""

    A = "A"
    B = "B"


@dataclass
class Base:
    """Base class."""

    tag: Tag


@dataclass
class A(Base):
    """Class A."""

    tag: Literal[Tag.A] = dataclasses.field(default=Tag.A, init=False)
    data: int


@dataclass
class B(Base):
    """Class B."""

    tag: Literal[Tag.B] = dataclasses.field(default=Tag.B, init=False)
    data: str


def do_something(x: Base) -> str:
    """Do something depending on the actual type of `x`."""
    if x.tag == Tag.A:
        return str(x.data + 1)
    if x.tag == Tag.B:
        return x.data + "a"
    typing.assert_never(x)
```
`mypy` finds the following "errors":
```
example.py:45: error: "Base" has no attribute "data"  [attr-defined]
example.py:47: error: "Base" has no attribute "data"  [attr-defined]
```
These errors disappear if `x` is of type `A | B` instead of `Base`.
```diff
-def do_something(x: Base) -> str:
+def do_something(x: A | B) -> str:
```

Reply:
Yes, you're correct—the base class doesn't seem necessary in this case.

Reply:
Fixed in d6f0a7c9acd299110087467a0b4de1771ba8e481 .